github/ratelimit.go

```diff
@@ -0,0 +1,219 @@
+package github
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"gopkg.in/src-d/go-log.v1"
+
+	"github.com/src-d/metadata-retrieval/utils"
+)
+
+var (
+	defaultAbuseRetryAfter = time.Minute
+)
+
+// RateLimitTransport implements GitHub GraphQL API v4 best practices for avoiding rate limits
+// https://developer.github.com/v4/guides/resource-limitations/#rate-limit
+// https://developer.github.com/v3/#abuse-rate-limits
+// RateLimitTransport will process a Request, and if the response could not be fetched
+// because of a RateLimit or an AbuseRateLimit, it will return an ErrorRateLimit
+// and it no longer process any further Requests until the Limit has been expired.
+// RateLimitTransport does not retry; that behaviour must be implemented by another Transport
+// Each client (with its own token) should use its own RateLimitTransport
+type RateLimitTransport struct {
+	transport   http.RoundTripper
+	lockedUntil time.Time
+	mutex       sync.Mutex
+	logger      log.Logger
+}
+
+// SetRateLimitTransport wraps the passed client.Transport with a RateLimitTransport
+func SetRateLimitTransport(client *http.Client, logger log.Logger) {
+	client.Transport = NewRateLimitTransport(client.Transport, logger)
+}
+
+// NewRateLimitTransport returns a new NewRateLimitTransport, who will call the passed
+// http.RoundTripper to process the http.Request
+// Each client (with its own token) should use its own RateLimitTransport
+func NewRateLimitTransport(rt http.RoundTripper, logger log.Logger) *RateLimitTransport {
+	return &RateLimitTransport{
+		transport: rt,
+		logger:    logger,
+	}
+}
+
+// RoundTrip executes a single HTTP transaction, returning a Response for the provided Request.
+// If the request hitted an API RateLimit or Abuse, it will return an ErrorRateLimit
+// and it no longer process any further Requests until the Limit has been expired.
+func (rt *RateLimitTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Make Requests serially
+	rt.lock()
+	defer rt.unlock()
+
+	if time.Now().Before(rt.lockedUntil) {
+		rt.logger.Infof("rate limit reached, sleeping until %s", rt.lockedUntil)
+		time.Sleep(rt.lockedUntil.Sub(time.Now()))
+	}
+
+	resp, err := rt.transport.RoundTrip(req)
+	if err != nil {
+		return resp, err
+	}
+
+	if errRateLimit := checkResponseRateLimit(resp, rt.logger); errRateLimit != nil {
+		rt.lockedUntil = errRateLimit.when()
+		rt.logger.Errorf(errRateLimit, "locking transport for %s", errRateLimit.when().Sub(time.Now()))
```

Comment:
most probably it should be warning as we expect hitting limit for big orgs and handle it correctly

Reply:
6aa93c50b76aea81e31d11f9d8a1fc2b36e10ee5 thanks